### PR TITLE
Elimination mode (with ,join command)

### LINKED ISSUE #410: Add elimination mode
Limit of how many lives one has, and if the user runs out, they're eliminated. Should work in single-player as well

### PR DESCRIPTION
**Pull request by taahamahdi**
Closes #410.

### CODE REVIEW

File: src/commands/game_commands/join.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export default class JoinCommand implements BaseCommand {
+    async call({ message, gameSessions }: CommandArgs) {
+        if (gameSessions[message.guildID] && gameSessions[message.guildID].eliminationMode && gameSessions[message.guildID].sessionInitialized) {
+            await sendErrorMessage(message, "Game already in session", "You can only join as a participant before an elimination game has started. Please wait until the current game ends.");
+        } else {
+            gameSessions[message.guildID].addParticipant(message.author);
+        }
+    }
+    aliases = ["j"];
+    help = {
```
Comment (Brainicism) [inline src/commands/game_commands/join.ts:13]: Don't document `,join`, it should only show up in the message that appears when you `,play elimination`

File: src/commands/game_commands/play.ts [hunk 2]
```diff
@@ -21,24 +28,65 @@ export default class PlayCommand implements BaseCommand {
             if (!voicePermissionsCheck(message)) {
                 return;
             }
+            const msgHasElimination = parsedMessage.components.length >= 1 && parsedMessage.components[0].toLowerCase() === "elimination";
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:31]: isEliminationMode?

File: src/commands/game_commands/play.ts [hunk 3]
```diff
@@ -21,24 +28,65 @@ export default class PlayCommand implements BaseCommand {
...
             if (!gameSessions[message.guildID]) {
                 const textChannel = message.channel;
-                const gameSession = new GameSession(textChannel, voiceChannel, message.author);
+                const gameOwner = message.author;
+                const lives = parsedMessage.components.length > 1 ? parseInt(parsedMessage.components[1], 10) : DEFAULT_LIVES;
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:35]: These elimination specific values should be instantiated if we're in elimination mode

File: src/models/elimination_scoreboard.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    decrementAllLives() {
+        for (const player of Object.values(this.players)) {
+            player.decrementLives();
+        }
+    }
+
+    gameFinished(): boolean {
+        const allEliminated = Object.values(this.players).every((player) => player.isEliminated());
```
Comment (Brainicism) [inline src/models/elimination_scoreboard.ts:64]: Add comments here to describe each case

File: src/models/elimination_scoreboard.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    getPlayerLives(userId: string): number {
+        return this.players[userId].getLives();
+    }
+
+    numberOfPlayers(): number {
+        return Object.keys(this.players).length;
+    }
+
+    getStartingLives(): number {
```
Comment (Brainicism) [inline src/models/elimination_scoreboard.ts:82]: Where is this used?

File: src/models/game_session.ts [hunk 6]
```diff
@@ -24,6 +25,10 @@ const LAST_PLAYED_SONG_QUEUE_SIZE = 10;
 export default class GameSession {
     private readonly startedAt: number;
 
+    public readonly eliminationMode: boolean;
```
Comment (Brainicism) [inline src/models/game_session.ts:28]: This should probably be of type `gameType`, the enum you created instead of specific boolean

File: src/models/game_session.ts [hunk 7]
```diff
@@ -291,6 +313,17 @@ export default class GameSession {
         this.guessTimeoutFunc = setTimeout(async () => {
             if (this.finished) return;
             logger.info(`${getDebugContext(message)} | Song finished without being guessed, timer of: ${time} seconds.`);
+            if (this.eliminationMode) {
+                const eliminationScoreboard = this.scoreboard as EliminationScoreboard;
+                eliminationScoreboard.decrementAllLives();
+                if (eliminationScoreboard.gameFinished()) {
+                    sendSongMessage(message, this.scoreboard, this.gameRound, true);
+                    await sendEndGameMessage({ channel: message.channel, authorId: message.author.id }, this);
+                    this.endRound(false);
```
Comment (Brainicism) [inline src/models/game_session.ts:322]: endSession already calls endRound

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/models/game_session.ts [hunk 8]
```diff
@@ -190,18 +210,15 @@ export default class GameSession {
             this.lastPlayedSongsQueue.shift();
         }
 
-        this.sessionInitialized = true;
         let randomSong: QueriedSong;
         try {
             randomSong = await selectRandomSong(guildPreference, this.lastPlayedSongsQueue);
             if (randomSong === null) {
-                this.sessionInitialized = false;
```
Comment (Brainicism) [inline src/models/game_session.ts:198]: why was this removed?

File: src/models/game_session.ts [hunk 9]
```diff
@@ -190,18 +210,15 @@ export default class GameSession {
...
                 sendErrorMessage(message, "Song Query Error", "Failed to find songs matching this criteria. Try to broaden your search.");
                 this.endSession();
                 return;
             }
         } catch (err) {
-            this.sessionInitialized = false;
```
Comment (Brainicism) [inline src/models/game_session.ts:204]: same here

File: src/models/game_session.ts [hunk 10]
```diff
@@ -217,7 +234,6 @@ export default class GameSession {
             await ensureVoiceConnection(this, state.client);
         } catch (err) {
             await this.endSession();
-            this.sessionInitialized = false;
```
Comment (Brainicism) [inline src/models/game_session.ts:220]: same here

File: src/models/game_session.ts [hunk 11]
```diff
@@ -338,4 +364,13 @@ export default class GameSession {
     resetLastPlayedSongsQueue() {
         this.lastPlayedSongsQueue = [];
     }
+
+    addParticipant(user: Eris.User) {
+        this.participants.add(user.id);
+        if (this.eliminationMode === GameType.ELIMINATION) {
+            const eliminationScoreboard = this.scoreboard as EliminationScoreboard;
+            eliminationScoreboard.addPlayer(user.id, getUserIdentifier(user), user.avatarURL);
+            this.scoreboard = eliminationScoreboard;
```
Comment (Brainicism) [inline src/models/game_session.ts:373]: is this needed?

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/models/game_session.ts [hunk 12]
```diff
@@ -166,17 +177,26 @@ export default class GameSession {
...
                 .increment("songs_guessed", 1);
-            if (!guildPreference.isGoalSet() || this.scoreboard.getWinners()[0].getScore() < guildPreference.getGoal()) {
-                this.startRound(guildPreference, message);
-            } else {
+            if (this.eliminationMode === GameType.ELIMINATION) {
+                const eliminationScoreboard = this.scoreboard as EliminationScoreboard;
+                if (eliminationScoreboard.gameFinished()) {
+                    logger.info(`${getDebugContext(message)} | Game session ended (one player alive in eliminationMode)`);
+                    await sendEndGameMessage({ channel: message.channel, authorId: message.author.id }, this);
+                    await this.endSession();
```
Comment (Brainicism) [inline src/models/game_session.ts:185]: Should there be a return here?

File: src/commands/game_commands/join.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+export default class JoinCommand implements BaseCommand {
+    async call({ message, gameSessions }: CommandArgs) {
+        if (!gameSessions[message.guildID] || gameSessions[message.guildID].eliminationMode === GameType.CLASSIC) {
+            return;
+        }
+        if (gameSessions[message.guildID].sessionInitialized) {
+            await sendErrorMessage(message, "Game already in session", "You can only join as a participant before an elimination game has started. Please wait until the current game ends.");
+        } else if (gameSessions[message.guildID].participants.has(message.author.id)) {
+            sendErrorMessage(message, "User already joined", `${bold(getUserIdentifier(message.author))} is already in the game.`);
```
Comment (Brainicism) [inline src/commands/game_commands/join.ts:14]: Pick either user or player, not both.

File: GAMEPLAY.md [hunk 14]
```diff
@@ -8,6 +8,9 @@ You can see the current scoreboard by typing `,score`. It is also shown after th
 
 See the latest updates to KMQ with `,news`. 
 
+## Game Modes
+See who can survive the longest into a KMQ game with elimination mode. Using `,play elimination x`, everyone starts with `x` lives and the last one alive wins. The first one to guess right keeps a life. Use elimination mode in conjunction with `,timer` to raise the pressure! Elimination mode and `,goal` are incompatible with each other.
```
Comment (Brainicism) [inline GAMEPLAY.md:12]: `The first one to guess right keeps a life` => `Guessing correctly will save your life, while everyone else loses one.`
`Elimination mode and ',goal' are incompatible with each other.` => ',goal' cannot be used in elimination mode

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/play.ts [hunk 15]
```diff
@@ -1,15 +1,23 @@
...
+    sendErrorMessage, getDebugContext, sendInfoMessage, getVoiceChannel, voicePermissionsCheck, getUserIdentifier,
 } from "../../helpers/discord_utils";
+import { deleteGameSession } from "../../helpers/management_utils";
 import { getGuildPreference } from "../../helpers/game_utils";
+import { bold } from "../../helpers/utils";
 import BaseCommand, { CommandArgs } from "../base_command";
 import _logger from "../../logger";
 
 const logger = _logger("play");
+const DEFAULT_LIVES = 3;
```
[the remark this answers is not in the document]
Comment (taahamahdi) [reply src/commands/game_commands/play.ts:12]: How's 5?

File: src/models/game_session.ts [hunk 16]
```diff
@@ -127,7 +130,9 @@ export default class GameSession {
 
     checkGuess(message: Eris.Message, modeType: string): number {
         if (!this.gameRound) return 0;
-        this.participants.add(message.author.id);
+        if (this.eliminationMode === GameType.ELIMINATION) {
```
[the remark this answers is not in the document]
Comment (taahamahdi) [reply src/models/game_session.ts:133]: Yup! I must've messed up during the switch away from the boolean eliminationMode, thanks for finding it!

File: src/models/game_session.ts (hunk 8 above)
Comment (taahamahdi) [reply src/models/game_session.ts:198]: My mistake, I wasn't sure where to set sessionInitialized back to false. Does it make sense to move it directly into the endSession method?

File: src/models/game_session.ts (hunk 11 above)
Comment (taahamahdi) [reply src/models/game_session.ts:373]: Yup, it's the only place where we can pass through the user's identifier and avatarURL to the elimination scoreboard (we only store users as their IDs in gameSession)

File: src/models/game_session.ts (hunk 12 above)
Comment (taahamahdi) [reply src/models/game_session.ts:185]: There should be, thanks :]

File: src/commands/game_commands/play.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (Brainicism) [reply src/commands/game_commands/play.ts:12]: Lives are an lower bound (when only one person guesses correctly until game ends) on how long a game session will last. Should be atleast 10 imo

[the remark this answers is not in the document]
Comment (taahamahdi) [reply src/commands/game_commands/play.ts:12]: Good point, I'll do 10

File: src/models/game_session.ts (hunk 8 above)
Comment (Brainicism) [reply src/models/game_session.ts:198]: nvm, this might be fine. just make sure typing `,play` multiple times in succession doesn't have any issues

File: src/models/game_session.ts (hunk 11 above)
Comment (Brainicism) [reply src/models/game_session.ts:373]: I meant re-assigning eliminationScoreboard to this.scoreboard here

File: src/models/game_session.ts (hunk 8 above)
Comment (taahamahdi) [reply src/models/game_session.ts:198]: I just tested ,play multiple times quickly and it behaved as expected (only one start message, the rest as "Game in session")

File: src/models/game_session.ts (hunk 11 above)
Comment (taahamahdi) [reply src/models/game_session.ts:373]: I had some confusion on how type assertion works, it isn't necessary since it isn't creating a new object

**Comment by taahamahdi** [comment]
The last thing I need to do is show the joined users in the `,join` command, then it's ready

File: src/commands/game_commands/begin.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+import { GameType } from "./play";
+import { getDebugContext, sendInfoMessage, getVoiceChannel } from "../../helpers/discord_utils";
+import _logger from "../../logger";
+
+const logger = _logger("begin");
+
+export default class BeginCommand implements BaseCommand {
+    async call({ message, gameSessions }: CommandArgs) {
+        const { guildID, author } = message;
+        if (!gameSessions[guildID] || gameSessions[message.guildID].gameType === GameType.CLASSIC || gameSessions[guildID].owner.id !== author.id) {
```
Comment (Brainicism) [inline src/commands/game_commands/begin.ts:12]: store gameSessions[guildID] in a variable and re-use it

File: src/commands/game_commands/join.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
+import BaseCommand, { CommandArgs } from "../base_command";
+import { GameType } from "./play";
+import { sendErrorMessage, sendInfoMessage, getUserIdentifier } from "../../helpers/discord_utils";
+import { bold } from "../../helpers/utils";
+
+export default class JoinCommand implements BaseCommand {
+    async call({ message, gameSessions }: CommandArgs) {
+        if (!gameSessions[message.guildID] || gameSessions[message.guildID].gameType === GameType.CLASSIC) {
```
Comment (Brainicism) [inline src/commands/game_commands/join.ts:8]: same here

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/join.ts [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+        if (!gameSession || gameSession.gameType === GameType.CLASSIC) {
+            return;
+        }
+        if (gameSession.sessionInitialized) {
+            await sendErrorMessage(message, "Game already in session", "You can only join as a participant before an elimination game has started. Please wait until the current game ends.");
+        } else if (gameSession.participants.has(message.author.id)) {
+            sendErrorMessage(message, "Player already joined", `${bold(getUserIdentifier(message.author))} is already in the game.`);
+        } else {
+            const previouslyJoinedPlayers = gameSession.scoreboard.getPlayerNames().reverse();
+            const players = `${bold(getUserIdentifier(message.author))}, ${previouslyJoinedPlayers.join(", ")}`.substring(0, 2048);
```
Comment (Brainicism) [inline src/commands/game_commands/join.ts:18]: Truncuate by number of players, not character limit. Let's do 10 players

File: src/commands/game_commands/join.ts [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+        const gameSession = gameSessions[message.guildID];
+        if (!gameSession || gameSession.gameType === GameType.CLASSIC) {
+            return;
+        }
+        if (gameSession.sessionInitialized) {
+            await sendErrorMessage(message, "Game already in session", "You can only join as a participant before an elimination game has started. Please wait until the current game ends.");
+        } else if (gameSession.participants.has(message.author.id)) {
+            sendErrorMessage(message, "Player already joined", `${bold(getUserIdentifier(message.author))} is already in the game.`);
+        } else {
+            const previouslyJoinedPlayers = gameSession.scoreboard.getPlayerNames().reverse();
```
Comment (Brainicism) [inline src/commands/game_commands/join.ts:17]: Why reverse?

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/join.ts (hunk 20 above)
Comment (taahamahdi) [reply src/commands/game_commands/join.ts:17]: I decided to show players in the order of most recently joined

File: src/commands/game_commands/join.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+            return;
+        }
+        if (gameSession.sessionInitialized) {
+            await sendErrorMessage(message, "Game already in session", "You can only join as a participant before an elimination game has started. Please wait until the current game ends.");
+            return;
+        }
+        if (gameSession.participants.has(message.author.id)) {
+            sendErrorMessage(message, "Player already joined", `${bold(getUserIdentifier(message.author))} is already in the game.`);
+        } else {
+            const previouslyJoinedPlayers = gameSession.scoreboard.getPlayerNames().reverse().slice(0, 10);
```
Comment (Brainicism) [inline src/commands/game_commands/join.ts:19]: Need to indicate that we've truncated the player list if we did

**Review by Brainicism — CHANGES_REQUESTED** [review]

**Review by Brainicism — APPROVED** [review]
